crates/doc/src/extractor.rs

```diff
@@ -0,0 +1,334 @@
+use crate::{compare::compare, ArchivedNode, AsNode, LazyNode, Node, Pointer};
+use bytes::BufMut;
+use std::borrow::Cow;
+use tuple::TuplePack;
+
+/// Extractor extracts locations from documents, and encapsulates various
+/// details of precisely how that's done.
+#[derive(Debug, Clone)]
+pub struct Extractor {
+    ptr: Pointer,
+    default: serde_json::Value,
+    is_uuid_v1_date_time: bool,
+}
+
+impl Extractor {
+    /// Build an Extractor for the JSON pointer.
+    /// If the location doesn't exist, then `null` is extracted.
+    pub fn new(ptr: &str) -> Self {
+        Self {
+            ptr: Pointer::from(ptr),
+            default: serde_json::Value::Null,
+            is_uuid_v1_date_time: false,
+        }
+    }
+
+    /// Build an extractor for the JSON pointer.
+    /// If the location doesn't exist, the provided value is extracted instead.
+    pub fn with_default(ptr: &str, default: serde_json::Value) -> Self {
+        Self {
+            ptr: Pointer::from(ptr),
+            default,
+            is_uuid_v1_date_time: false,
+        }
+    }
+
+    /// Build an extractor for the JSON pointer, which is a v1 UUID.
+    pub fn for_uuid_v1_date_time(ptr: &str) -> Self {
+        Self {
+            ptr: Pointer::from(ptr),
+            default: serde_json::Value::Null,
+            is_uuid_v1_date_time: true,
+        }
+    }
+
+    /// Query the value extracted from the document.
+    /// The result is Ok if a literal document node is extracted,
+    /// or Err if the document value doesn't exist and an alternative
+    /// serde_json::Value was extracted.
+    ///
+    /// Most commonly this is a default value for the location
+    /// and a borrowed Value is returned.
+    ///
+    /// Or, it may be a dynamic Value extracted from a UUID timestamp.
+    pub fn query<'s, 'n, N: AsNode>(
+        &'s self,
+        doc: &'n N,
+    ) -> Result<&'n N, Cow<'s, serde_json::Value>> {
+        let Some(node) = self.ptr.query(doc) else {
+            return Err(Cow::Borrowed(&self.default));
+        };
+
+        if self.is_uuid_v1_date_time {
+            if let Some(date_time) = match node.as_node() {
+                Node::String(s) => Some(s),
+                _ => None,
+            }
+            .and_then(|s| uuid::Uuid::parse_str(s).ok())
+            .and_then(|u| u.get_timestamp())
+            .and_then(|t| {
+                let (seconds, nanos) = t.to_unix();
+                time::OffsetDateTime::from_unix_timestamp_nanos(
+                    seconds as i128 * 1_000_000_000 + nanos as i128,
+                )
+                .ok()
+            }) {
+                return Err(Cow::Owned(serde_json::Value::String(
```

Comment:
It certainly could be a new enum, but using `Result` to represent a found / not-found result is in keeping with how the standard library uses it.  For example https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search

In this sense the `Result` built-ins are nice because you can do things like `result.ok()` to discard an alternative value and use it strictly as an `Option` over "found" status.

Reply:
But unlike binary_search, this function always returns "something", specifically something that has an implementation of `AsNode`.

Reply:
I'm losing the thread here. `query(..)` returns an extracted value, which can either be a literal document node (`Ok`) or an alternative value (`Err`).